src/python/pants/backend/audit/audit.py

```diff
@@ -0,0 +1,131 @@
+# Copyright 2024 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass
+from typing import Any, ClassVar, Generic, Iterable, TypeVar
+
+from pants.core.goals.lint import REPORT_DIR as REPORT_DIR  # noqa: F401
+from pants.core.goals.multi_tool_goal_helper import determine_specified_tool_ids
+from pants.engine.collection import Collection
+from pants.engine.console import Console
+from pants.engine.engine_aware import EngineAwareParameter, EngineAwareReturnType
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.rules import collect_rules, goal_rule
+from pants.engine.target import FieldSet, FilteredTargets
+from pants.engine.unions import UnionMembership, union
+
+logger = logging.getLogger(__name__)
+_FS = TypeVar("_FS", bound=FieldSet)
+
+
+@dataclass(frozen=True)
+class AuditResult:
+    resolve_name: str
+    lockfile: str
+    report: str
+
+
+@dataclass(unsafe_hash=True)
+class AuditResults(EngineAwareReturnType):
+    """Zero or more AuditResult objects for a single auditor."""
+
+    results: tuple[AuditResult, ...]
+    auditor_name: str
+
+    def __init__(self, results: Iterable[AuditResult], *, auditor_name: str) -> None:
+        self.results = tuple(results)
+        self.auditor_name = auditor_name
+
+    def cacheable(self) -> bool:
+        """Is marked uncacheable to ensure that it always renders."""
+        return False
+
+
+@union
+@dataclass(unsafe_hash=True)
+class AuditRequest(Generic[_FS], EngineAwareParameter):
+    """A union for targets that should be audited.
+
+    Subclass and install a member of this type to provide an auditor.
+    """
+
+    field_set_type: ClassVar[type[_FS]]  # type: ignore[misc]
+    tool_id: ClassVar[str]
+
+    field_sets: Collection[_FS]
+
+    def __init__(self, field_sets: Iterable[_FS]) -> None:
+        self.field_sets = Collection[_FS](field_sets)
+
+    def debug_hint(self) -> str:
+        return self.tool_id
+
+    def metadata(self) -> dict[str, Any]:
+        return {"addresses": [fs.address.spec for fs in self.field_sets]}
+
+
+class AuditSubsystem(GoalSubsystem):
+    name = "audit"
```

Comment:
Should this be `experimental-audit` for now like we did for `deploy`?

Reply:
We certainly could. I guess we could look at how big risk for change is there?
Deploy feels like it could be more sensitive to how it works, compared to presenting a list of applicable security reports, which could warrant this to go straight to a "stable" goal name.. for me it's fine either way, but I'm 👍🏽 for having the discussion to settle it.